packages/react-aria-components/src/GridList.tsx

```diff
@@ -0,0 +1,135 @@
+import React, {createContext, ForwardedRef, forwardRef, useContext} from 'react';
+import {TextContext} from './Text';
+
+export interface GridListProps<T> extends Omit<AriaGridListProps<T>, 'children'>, CollectionProps<T>, StyleProps, SlotProps {}
+
+export const GridListContext = createContext<ContextValue<GridListProps<any>, HTMLUListElement>>(null);
+const InternalGridListContext = createContext<ListState<unknown>>(null);
+
+function GridList<T extends object>(props: GridListProps<T>, ref: ForwardedRef<HTMLUListElement>) {
+  [props, ref] = useContextProps(props, ref, GridListContext);
+  let {portal, collection} = useCollection(props);
+  let state = useListState({
+    ...props,
+    collection,
+    children: null
+  });
+
+  let {gridProps} = useGridList(props, state, ref);
+
+  let children = useCachedChildren({
+    items: collection,
+    children: (item: Node<T>) => {
+      switch (item.type) {
+        case 'item':
+          return <GridListItem item={item} />;
+        default:
+          throw new Error('Unsupported node type in GridList: ' + item.type);
+      }
+    }
+  });
+
+  return (
+    <ul
+      {...filterDOMProps(props)}
+      {...gridProps}
+      ref={ref}
+      slot={props.slot}
+      style={props.style}
+      className={props.className ?? 'react-aria-GridList'}>
+      <InternalGridListContext.Provider value={state}>
+        {children}
+      </InternalGridListContext.Provider>
+      {portal}
+    </ul>
+  );
+}
+
+/**
+ * A grid list displays a list of interactive items, with support for keyboard navigation,
+ * single or multiple selection, and row actions.
+ */
+const _GridList = forwardRef(GridList);
+export {_GridList as GridList};
+
+function GridListItem({item}) {
+  let state = useContext(InternalGridListContext);
+  let ref = React.useRef();
+  let {rowProps, gridCellProps, descriptionProps, ...states} = useGridListItem(
+    {node: item},
+    state,
+    ref
+  );
+
+  let {hoverProps, isHovered} = useHover({
+    isDisabled: !states.allowsSelection && !states.hasAction
+  });
+
+  let {isFocusVisible, focusProps} = useFocusRing();
+  let {checkboxProps} = useGridListSelectionCheckbox(
+    {key: item.key},
+    state
+  );
+
+  let props: ItemProps<unknown> = item.props;
+  let renderProps = useRenderProps({
+    ...props,
+    children: item.rendered,
+    defaultClassName: 'react-aria-Item',
+    values: {
+      ...states,
+      isHovered,
+      isFocusVisible,
+      selectionMode: state.selectionManager.selectionMode,
+      selectionBehavior: state.selectionManager.selectionBehavior
+    }
+  });
+
+  return (
+    <li
+      {...mergeProps(rowProps, focusProps, hoverProps)}
+      {...renderProps}
+      ref={ref}
+      data-hovered={isHovered || undefined}
+      data-focused={states.isFocused || undefined}
+      data-focus-visible={isFocusVisible || undefined}
+      data-pressed={states.isPressed || undefined}>
+      {/* TODO: hard to style this element... */}
```

Comment:
Yeah. In most cases I try to keep the components mapping 1:1 with DOM nodes, which allows the user to easily add custom classes, data attrs, etc. There are a few cases so far where that's not true because accessibility requires some extra DOM node, and it would be annoying to require writing a whole separate element in apps. This is one of those cases. I'm not totally sure what to do about it. For styling with CSS, it's easy enough to target if you need to (via the role attribute or something), but you can't easily customize it otherwise...